### PR TITLE
Fix account decoding

### PR DESCRIPTION
**Pull request by AndreaLanfranchi**
- fix decode account from encoded rlp : 1) Check payload is either empty or > 1 (otherwise first `encoded_payload[pos++]` gives undefined behavior); 2) remove duplicate checks and reduce to a loop; 
- fix decode incarnation from encoded rlp : implementation did not include all checks of the previous
- move `decode_account_from_storage` to `Account::from_encoded_storage`
- move `extract_incarnation` to `Account::incarnation_from_encoded_storage`
- addedd missing tests for wrong payloads

### CODE REVIEW

File: core/silkworm/types/account.cpp [hunk 1]
```diff
@@ -84,85 +84,96 @@ size_t Account::encoding_length_for_storage() const {
...
-}
-
-std::pair<Account, rlp::DecodingResult> decode_account_from_storage(ByteView encoded) noexcept {
+std::pair<Account, rlp::DecodingResult> Account::from_encoded_storage(ByteView encoded_payload) noexcept {
     Account a{};
-    if (encoded.empty()) {
+    if (encoded_payload.empty()) {
         return {a, rlp::DecodingResult::kOk};
+    } else if (encoded_payload.length() == 1) {
+        // Must be at least 2 bytes : field_set + len of payload
```
Comment (yperbasis) [inline core/silkworm/types/account.cpp:92]: That's not right: `encode_for_storage` on an empty account will return a string of length 1 with a single 0 byte.

Comment (AndreaLanfranchi) [reply core/silkworm/types/account.cpp:92]: Ok skimmed over it. The check has to be hardened so if `encoded_payload[0] != 0` then there MUST be something afterwards. Doing now

Comment (yperbasis) [reply core/silkworm/types/account.cpp:92]: The simplest thing is to remove this check altogether. The length is already checked in the main loop.

Comment (AndreaLanfranchi) [reply core/silkworm/types/account.cpp]: Done 8425b9b

File: core/silkworm/types/account.cpp [hunk 2]
```diff
@@ -84,85 +84,109 @@ size_t Account::encoding_length_for_storage() const {
...
+                case 1: {
+                    const std::optional<uint64_t> nonce{
+                        endian::from_big_compact<uint64_t>(encoded_payload.substr(pos, len))};
+                    if (nonce == std::nullopt) {
+                        return {a, rlp::DecodingResult::kLeadingZero};
+                    }
+                    a.nonce = *nonce;
+                } break;
+                case 2:
+                    std::memcpy(&as_bytes(a.balance)[32 - len], &encoded_payload[pos], len);
```
Comment (yperbasis) [inline core/silkworm/types/account.cpp:129]: Let's use `endian::from_big_compact<intx::uint256>` for consistency

Comment (AndreaLanfranchi) [reply core/silkworm/types/account.cpp:129]: Sure ... I have question though.
Actually `from_big_compact` returns std::nullopt in two cases:

- either size of target is < size of data string (overflow)
- or there are leading zeroes and are not allowed

How we could efficiently make a distinction ?
Wouldn't it be better to transform `from_big_compact` into a parser into the rlp namespace ?
E.g.

```
template <typename UnsignedInteger>
static rlp::DecodingResult parse_big_compact(ByteView data, UnsignedInteger& target, bool allow_leading_zeros = false) {
 [...]
 target = be:load(x)
 return rlp::DecodingResult::kOk
)
```

Comment (yperbasis) [reply core/silkworm/types/account.cpp]: Perhaps. I don't have a strong view on that. An argument against moving `parse_big_compact` into RLP is that it's not particularly RLP-specific, but rather a useful low-level function. If you decide to move it into RLP, please move it alongside `to_big_compact` and in a separate PR.

**Review by yperbasis — APPROVED** [review]

**Pull request merged**